src/decomp_2d.f90

```diff
@@ -414,43 +419,35 @@ subroutine decomp_info_init(nx, ny, nz, decomp)
 #endif
 
       ! check if additional memory is required
-      ! *** TODO: consider how to share the real/complex buffers
       if (buf_size > decomp_buf_size) then
          decomp_buf_size = buf_size
 #if defined(_GPU)
          call decomp_2d_cumpi_init(buf_size)
 #endif
-         if (allocated(work1_r)) deallocate (work1_r)
-         if (allocated(work2_r)) deallocate (work2_r)
-         if (allocated(work1_c)) deallocate (work1_c)
-         if (allocated(work2_c)) deallocate (work2_c)
-         allocate (work1_r(buf_size), STAT=status)
-         if (status /= 0) then
-            errorcode = 2
-            call decomp_2d_abort(__FILE__, __LINE__, errorcode, &
-                                 'Out of memory when allocating 2DECOMP workspace')
-         end if
-         allocate (work2_r(buf_size), STAT=status)
+         if (associated(work1_r)) nullify (work1_r)
+         if (associated(work2_r)) nullify (work2_r)
+         if (associated(work1_c)) nullify (work1_c)
+         if (associated(work2_c)) nullify (work2_c)
+         if (allocated(work1)) deallocate (work1)
+         if (allocated(work2)) deallocate (work2)
+         allocate (work1(2 * buf_size), STAT=status)
          if (status /= 0) then
             errorcode = 2
             call decomp_2d_abort(__FILE__, __LINE__, errorcode, &
                                  'Out of memory when allocating 2DECOMP workspace')
          end if
-         allocate (work1_c(buf_size), STAT=status)
-         if (status /= 0) then
-            errorcode = 2
-            call decomp_2d_abort(__FILE__, __LINE__, errorcode, &
-                                 'Out of memory when allocating 2DECOMP workspace')
-         end if
-         allocate (work2_c(buf_size), STAT=status)
+         allocate (work2(2 * buf_size), STAT=status)
          if (status /= 0) then
             errorcode = 2
             call decomp_2d_abort(__FILE__, __LINE__, errorcode, &
                                  'Out of memory when allocating 2DECOMP workspace')
          end if
+         call c_f_pointer(c_loc(work1), work1_r, [2 * buf_size])
+         call c_f_pointer(c_loc(work2), work2_r, [2 * buf_size])
```

Comment:
Could we use `[buf_size]` here as the `shape` too? This should (I think) prevent the `real` pointer from accessing beyond the "end" of the array.

Reply:
This modification should not hurt

Reply:
Solved in https://github.com/2decomp-fft/2decomp-fft/pull/218/commits/cc648d9219ab73c335bd97670e330c3c3277d41c